src/calcium/CalciumTypes.jl

```diff
@@ -76,4 +96,111 @@ function _qqbar_clear_fn(a::qqbar)
    ccall((:qqbar_clear, libcalcium), Nothing, (Ref{qqbar},), a)
 end
 
+################################################################################
+#
+#  Types and memory management for CalciumField
+#
+################################################################################
+
+ca_ctx_options = [
+    :verbose,
+    :print_flags,
+    :mpoly_ord,
+    :prec_limit,
+    :qqbar_deg_limit,
+    :low_prec,
+    :smooth_limit,
+    :lll_prec,
+    :pow_limit,
+    :use_gb,
+    :gb_length_limit,
+    :gb_poly_length_limit,
+    :gb_poly_bits_limit,
+    :vieta_limit,
+    :trig_form]
+
+mutable struct CalciumField <: Field
+   ext_cache_items::Ptr{Nothing}
+   ext_cache_length::Int
+   ext_cache_alloc::Int
+   ext_cache_hash_size::Int
+   ext_cache_hash_table::Ptr{Nothing}
+   field_cache_items::Ptr{Nothing}
+   field_cache_length::Int
+   field_cache_alloc::Int
+   field_cache_hash_size::Int
+   field_cache_hash_table::Ptr{Nothing}
+   field_qq::Ptr{Nothing}
+   field_qq_i::Ptr{Nothing}
+   mctx::Ptr{Nothing}
+   mctx_len::Int
+   options::Ptr{Int}
+   # end C struct
+
+   extended::Bool
+   refcount::Int
+
+   function CalciumField(; extended::Bool=false, options::Dict{Symbol,Int}=Dict{Symbol,Int}())
+      C = new()
+
+      ccall((:ca_ctx_init, libcalcium), Nothing, (Ref{CalciumField}, ), C)
+      finalizer(_CalciumField_clear_fn, C)
+      C.extended = extended
+
+      for (opt, value) in options
+         i = findfirst(isequal(opt), ca_ctx_options)
+         (i == nothing) && error("unknown option ", opt)
+         ccall((:ca_ctx_set_option, libcalcium), Nothing, (Ref{CalciumField}, Int, Int), C, i - 1, value)
+      end
+
+      C.refcount = 1
+      return C
+   end
+end
+
+function options(C::CalciumField)
+   d = Dict{Symbol,Int}()
+   for i=1:length(ca_ctx_options)
+      d[ca_ctx_options[i]] = ccall((:ca_ctx_get_option, libcalcium), Int, (Ref{CalciumField}, Int), C, i - 1)
+   end
+   return d
+end
+
+function _CalciumField_clear_fn(C::CalciumField)
+   C.refcount -= 1
+   if C.refcount == 0
+      ccall((:ca_ctx_clear, libcalcium), Nothing, (Ref{CalciumField},), C)
+   end
+end
+
+mutable struct ca <: FieldElem
+   field::Int
+   data0::UInt
+   data1::UInt
+   data2::UInt
+   data3::UInt
+   # end C struct
+
+   parent::CalciumField
+
+   function ca(ctx::CalciumField)
+      z = new()
+      ccall((:ca_init, libcalcium), Nothing,
+                (Ref{ca}, Ref{CalciumField}), z, ctx)
+      z.parent = ctx
+      z.parent.refcount += 1
+      finalizer(_ca_clear_fn, z)
+      return z
+   end
+
+end
+
+function _ca_clear_fn(a::ca)
+   ccall((:ca_clear, libcalcium),
+        Nothing, (Ref{ca}, Ref{CalciumField}), a, a.parent)
+   a.parent.refcount -= 1
+   if a.parent.refcount == 0
+      ccall((:ca_ctx_clear, libcalcium), Nothing, (Ref{CalciumField},), a.parent)
```

Comment:
Codecov seems to indicate ca_ctx_clear is never called. Could it be the refcounting is not working?

Reply:
Yeah, this is precisely the path that only gets called if Julia calls finalizers out of order, which only happens on some versions and on some machines. However, introducing a helper function should solve the coverage issue.